parser/typecheck/typecheck.py

```diff
@@ -0,0 +1,227 @@
+    """
+
+
+@dataclass
+class FunctionType(TypeInfo):
+    arg_types: list[TypeInfo]
+    ret_type: TypeInfo
+
+
+@dataclass
+class NameInfo:
+    decl_scope: Scope
+    ident: str
+    tp_info: TypeInfo
+    # node: AstNode  # <-- Why do we need this?
+    is_param: bool = field(default=False, kw_only=True)
+
+
+@dataclass
+class FuncInfo(NameInfo):
+    tp_info: FunctionType  # Overrides types (doesn't change order)
+    params_info: list[ParamInfo]
+    # Can't just pass default_factory=Scope as it is only defined below
+    subscope: Scope = field(default_factory=lambda: Scope())
+
+    @classmethod
+    def from_param_info(
+            cls, decl_scope: Scope, ident: str, params_info: list[ParamInfo],
+            ret_type: TypeInfo, subscope: Scope = None):
+        subscope = subscope or Scope()
+        tp_info = FunctionType([p.tp for p in params_info], ret_type)
+        return cls(decl_scope, ident, tp_info, params_info, subscope)
+
+
+@dataclass
+class ParamInfo:
+    name: str
+    tp: TypeInfo
+
+
+@dataclass
+class Scope:
+    declared: dict[str, NameInfo] = field(default_factory=dict)
+    used: dict[str, NameInfo] = field(default_factory=dict)
+    # Add references to outer scopes' variables that we use
+
+
+class NameResolutionError(BaseLocatedError):
+    pass
+
+
+# The reason `let` isn't used is because we don't want to imply similarity
+#   between parameters as local variables (where none exists in Scratch).
+#   Also, we might want to use `let` later as a modifier to bind it to
+#     an actual local var.
+# Don't need to `sys.intern` these manually as Python automatically does
+#   this for literals.
+PARAM_TYPES = {'number', 'string', 'val', 'bool'}
+
+
+# Variables:
+#  - We can prevent usages before the variable is declared in 2 ways:
+#    - Based on time: very sensible, like JS, but requires too many runtime features
+#    - Based on location: somewhat makes sense except for inner functions -
+#       they may be called later so should be able to access any variables.
+#  - Or we can just ignore it (e.g. `var` in JS) and pretend everything was
+#     declared at the top (but not assigned to - i.e. hoist `var foo;` to top).
+# To minimise accidental errors, option 1.2 is best
+#  (errors shouldn't pass silently, and that method requires no special runtime)
+class NameResolver:
+    def __init__(self, astgen: AstGen):
+        self.astgen = astgen
+        self.src = self.astgen.src
+        self.top_scope: Scope | None = None
+
+    def _init(self):
+        self.ast = self.astgen.parse()
+        self.top_scope = Scope()
+
+    def run(self):
+        if self.top_scope:
+            return self.top_scope
+        self._init()
+        self.run_on_new_scope(self.ast.statements, curr_scope=self.top_scope)
+        return self.top_scope
+
+    def run_on_new_scope(self, block: list[AstNode], scope_stack: list[Scope] = None,
+                         curr_scope: Scope = None):
```

Comment:
Maybe do `parent_scopes` instead of `scope_stack` to make it clear that the current one isn't in it.